### PR TITLE
Dynamic Images

### PR DESCRIPTION
**Pull request by HardeepAsrani**
Closes https://github.com/Codeinwp/otter-blocks/issues/996.

### Screenshots 

Screenshot 2022-07-19 at 3 00 00 PM

Screenshot 2022-07-19 at 3 04 47 PM

----

### Test instructions

---- 

### Checklist before the final review

- [ ] Visual elements are not affected by independent changes.
- [ ] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [ ] It loads additional script in frontend only if it is required.
- [ ] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [ ] In case of deprecation, old blocks are safely migrated.
- [ ] It is usable in Widgets and FSE.

### CODE REVIEW

File: src/blocks/plugins/dynamic-content/media/media-content.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+	}, [ selected ]);
+
+	const changeAttributes = obj => {
+		let attrs = { ...attributes };
+
+		Object.keys( obj ).forEach( o => {
+			attrs[ o ] = obj[ o ];
+		});
+
+		attrs = Object.fromEntries( Object.entries( attrs ).filter( ([ _, v ]) => ( null !== v && '' !== v ) ) );
```
Comment (Soare-Robert-Daniel) [inline src/blocks/plugins/dynamic-content/media/media-content.js:181]: You can simplify this part to `const attrs = pickBy({...attributes, ...obj}, v => v);`

**Comment by HardeepAsrani** [comment]
@mghenciu @selul I had one big pain point in Dynamic Images (of course, it's the media modal): https://www.loom.com/share/7ad5904ed15a41f0a7b13f19af2dd891 - wanted to have your thoughts on it.

**Comment by HardeepAsrani** [comment]
An update: I've found one way to overcome this. It will be by assigning a dummy ID (which is saved in attributes) to each selected image (probably 8 digits so as not to conflict with core post/image IDs) and also having the same attribute as a query parameter in the image URL and combining it with https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/media-upload/README.md

It might have issues in some places if the block tries to do something specific with the image ID, but it should cover most cases.

So theoretically, it's possible to make this work. I just need to put the entire thing in code now. Goddamn it, they redid half the WordPress in React yet the media modal is still Backbone.

**Comment by mghenciu** [comment]
A few things I had in mind when we started working on this:
- how would the AFC work with images, for example if I create multiple ACF fields with images, will those appear in the media library, or how would I select a Image field defined in AFC (or maybe multiple images on a page)
- we need to make sure **we don’t reset the Image Settings**, when changing the dynamic source (cover mode, overlay, etc). So I am thinking about this just as an image source configuration, all the other image settings should remain untouched. 
- Woocommerce. When this is done, a walkthrough in the dynamic library would help.
- I think Marius mentioned this already. At this point is not clear that the Dynamic Content (in gallery) is coming from Otter, nor when hovering clicking on the Locked ones, no feedback about upgrading is shown.
Maybe the Otter Icon in the Tabs would help, and a Tooltip for the Locked Pro features.

**Comment by HardeepAsrani** [comment]
@mghenciu Things should be more clear today or tomorrow when everything is done then we can go through and make sure everything is as we'd expect them to be. I'm sure there will be many bugs with this, it's been pretty shitty coding the media library part of it.

**Comment by HardeepAsrani** [comment]
Feedback appreciated @Codeinwp/design-team

**Comment by JohnPixle** [comment]
Looks good @HardeepAsrani I just tested it.

Here is a 2-min video: https://www.loom.com/share/a257ecfce5d2481686e96f80b64fab0c

## Feedback

### Display rendered image before inserting
- [ ] I was thinking that if possible we should show the actual dynamic image, rendered in the popup before it gets inserted into the layout. Not sure if it's technically an option.

### Fallback image
- [ ] Not sure how friendly this is, but I don't have any alternatives to suggest for now :)

Great work, looking forward to using this 👍🏻

**Comment by mghenciu** [comment]
I agree with what John mentioned, **about the preview,** that will really help.
Probably for the **Fallback image**, we can have a placeholder in the input field: _Paste the Image URL_.

___
There look to be some bugs, for example with the Media & Text block, I can't access anymore the Dynamic Content Tab, after the first use. So probably a proper testing is needed, with all the blocks that use this option.

https://user-images.githubusercontent.com/52494172/179744991-5b547122-0741-406f-bd9c-a69dee5ed9af.mov

 ___
I would still apreciate a walkthrough through the dynamic library, in the next period. As not all the options are clear for me. For example I am not sure how Meta can be used.
Essentially the idea would be to have at least a few **Jobs to be done** that users may want, and we test/make sure those work as expected. 
### A few examples of Jobs to be done:
- building a dynamic staff directory
- Dynamic image for single posts recipes (recipe site)
- dynamic Testimonials section
- Dynamic books listing, where each book has a page (not necessarily for sale, just for presentation purposes)

**Comment by HardeepAsrani** [comment]
>  I was thinking that if possible we should show the actual dynamic image, rendered in the popup before it gets inserted into the layout. Not sure if it's technically an option.

It will be a pain in the ass to show the actual preview + it will be useless if the most is not published as we can't get the info such as featured image and etc unless the most is published, so it will ever be a half baked option.

**Comment by HardeepAsrani** [comment]
@mghenciu Yea, it seems it there might be some blocks where we will have to test. Hopefully, we will know all once @irinelenache has done the QA.

Right now I can for sure confirm it won't work with Core's Query block in FSE as the query block gives us no way to get the ID of the post.

**Comment by HardeepAsrani** [comment]
Also for meta option, it's for Custom Field. You can get any image value from Custom Fields.

**Comment by JohnPixle** [comment]
That's ok by me then, let's see how we do with the testing 👍🏻

**Comment by HardeepAsrani** [comment]
@irinelenache These should be fixed. You can take another look to confirm. Also, maybe keep an eye out on debug.log to make sure we don't see any debugging errors.

**Comment by irinelenache** [comment]
@HardeepAsrani The mentioned issues are fixed, thank you 🚀 

Tested the **Section bg** and it was working fine before these fixes, now it doesn't work at all. I added something both for **Section bg** and **Section column bg** and it's working fine in editor only
editor: https://vertis.d.pr/4ZEeJZ
frontend: https://vertis.d.pr/zEFIDl

You can check the issue on this instance:
```
 Admin area URL: https://dynamic-image.s3-tastewp.com/2022/07/27/section-bg-section-column-bg/
 Username: irinel 
 Password: Z-f7V6LRh8E 
```

**Comment by HardeepAsrani** [comment]
@irinelenache It should be fixed now. Let me know if you find anything else or we can merge this!

**Comment by irinelenache** [comment]
@HardeepAsrani Checked again and now everything works fine, will move it to **Ready to merge** 🚀

**Comment by HardeepAsrani** [comment]
Thank you @irinelenache!

**Pull request merged**